src/test/conditions/DerChoice.h

```diff
@@ -0,0 +1,478 @@
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_TEST_CONDITIONS_DERCHOICE_H_INCLUDED
+#define RIPPLE_TEST_CONDITIONS_DERCHOICE_H_INCLUDED
+
+#include <ripple/conditions/impl/Der.h>
+
+#include <algorithm>
+#include <fstream>
+#include <string>
+
+#include <boost/filesystem.hpp>
+
+namespace ripple {
+
+namespace test {
+
+struct DerChoiceBaseClass
+{
+    virtual ~DerChoiceBaseClass() = default;
+
+    virtual std::uint8_t
+    type() const = 0;
+
+    virtual
+    void
+    encode(cryptoconditions::der::Encoder& s) const = 0;
+
+    virtual
+    void
+    decode(cryptoconditions::der::Decoder& s) = 0;
+
+    virtual int
+    compare(
+        DerChoiceBaseClass const& rhs,
+        cryptoconditions::der::TraitsCache& traitsCache) const = 0;
+
+    virtual
+    std::uint64_t
+    derEncodedLength(
+        boost::optional<cryptoconditions::der::GroupType> const& parentGroupType,
+        cryptoconditions::der::TagMode encoderTagMode,
+        cryptoconditions::der::TraitsCache& traitsCache) const = 0;
+
+    // for debugging
+    virtual
+    void
+    print(std::ostream& ostr, bool ordered=false) const = 0;
+};
+
+bool
+equal(DerChoiceBaseClass const* lhs, DerChoiceBaseClass const* rhs);
+
+inline
+bool
+equal(
+    std::unique_ptr<DerChoiceBaseClass> const& lhs,
+    std::unique_ptr<DerChoiceBaseClass> const& rhs)
+{
+    return equal(lhs.get(), rhs.get());
+}
+
+inline
+bool
+operator==(
+    std::unique_ptr<DerChoiceBaseClass> const& lhs,
+    std::unique_ptr<DerChoiceBaseClass> const& rhs)
+{
+    return equal(lhs, rhs);
+}
+
+inline
+bool
+operator!=(
+    std::unique_ptr<DerChoiceBaseClass> const& lhs,
+    std::unique_ptr<DerChoiceBaseClass> const& rhs)
+{
+    return !operator==(lhs, rhs);
+}
+
+struct DerChoiceDerived1 : DerChoiceBaseClass
+{
+    Buffer buf_;
+    std::vector<std::unique_ptr<DerChoiceBaseClass>> subChoices_;
+    std::int32_t signedInt_;
+
+    DerChoiceDerived1() = default;
+
+    DerChoiceDerived1(
+        std::vector<char> const& b,
+        std::vector<std::unique_ptr<DerChoiceBaseClass>> sub,
+        std::int32_t si);
+
+    template<class F>
+    void withTuple(F&& f, cryptoconditions::der::TraitsCache& traitsCache);
+
+    template<class F>
+    void withTuple(F&& f, cryptoconditions::der::TraitsCache& traitsCache) const
+    {
+        const_cast<DerChoiceDerived1*>(this)->withTuple(std::forward<F>(f), traitsCache);
+    }
```

Comment:
Fixed in `a028ad7 [fold] Responding to Ed PR feedback 1`

And an explanation: without the `const_cast` we have to duplicate all the `withTuple` functions, and I decided `const_cast` was better than duplication. However, given your "eeek!" I'm OK just duplicating the code.